### PR TITLE
WIP: stats: read and modify stats only when holding the cache lock

### PR DESCRIPTION
**Pull request by timblechmann**
* untested
* we might want to separate locks for stats and cache

### CODE REVIEW

File: clcache.py [hunk 1]
```diff
@@ -148,29 +148,28 @@ def cacheDirectory(self):
         return self.dir
 
     def clean(self, stats, maximumSize):
-        with self.lock:
-            currentSize = stats.currentCacheSize()
-            if currentSize < maximumSize:
-                return
+        currentSize = stats.currentCacheSize()
```
Comment (webmaster128) [inline clcache.py]: Why did you pull this outside the lock? currentCacheSize() is a read operation that should be protected.

Comment (timblechmann) [inline clcache.py]: `clean` is called under the cache lock (i've guarded each top-level instant of `stats` with the lock)

**Comment by webmaster128** [comment]
I don't think that this is going to end up in a robust solution as it is now because every caller has to take care to properly lock the stats in places where you want to have a higher level of abstraction.

Having a single lock for stats and cache is very problematic because you'll end up with 7 waiting cores that just need to read or write a tiny bit from/to stats.txt because 1 core is writing a big object file on a 8 core machine. Even in not so bad cases, waiting will sum up easily.

As @frerich suggested in #70, we should use CacheStatistics's methods to lock something.

What I have in mind is the following:

* Get rid of CacheStatistics.save()
* Create a separate lock for stats.txt: "CacheStatisticsLock" like ObjectCacheLock width other mutex name
* Add a generic private getter and setter `_get(key)` and `_put(key, newValue)` that *lock, read/write file, unlock*
* Make existing getters use the new getter, e.g.
```
    def numCallsWithoutSourceFile(self):
        return self._get("CallsWithoutSourceFile")
```
* Make existing setters use the new setter, e.g.

```
    def registerCallWithMultipleSourceFiles(self):
        old = self._get("CallsWithMultipleSourceFiles")
        self._put("CallsWithMultipleSourceFiles", old+1)
```

* Now we don't want 2 locks for increase operation, so add `_increase(key, ammount)` to `_get` and `_put` that does *lock, read, increase, write, unlock* and refactor the method above to

```
    def registerCallWithMultipleSourceFiles(self):
        self._increase("CallsWithMultipleSourceFiles", 1)
```

* Since every data comes from the stats.txt file now, we can remove the in memory copy of values `self._stats`

This reduces the locking time to a minimal, keeps the code simple because only 3 methods have locking logic in it, solves #70 because we don't need to keep a copy of the current stats value in memory, can be improved by using SQLite later on.

**Comment by frerich** [comment]
Thanks for the patch! Except for minor cosmetic issues (you introduced spaces within `()` in a couple of places, e.g. `CacheStatistics(cache)` became `CacheStatistics( cache )` in a couple of places), I couldn't spot any obvious problems, i.e. I suspect it does the Right Thing(tm).

At the same time, that's my main concern with the patch. I find it hard to tell whether it's correct, which is not you to blame but rather a problem with the current code base. Hence, I'd feel a bit better if we could change the code such that instead of having no obvious bugs (with respect to concurrency), it obviously has no bugs.

My idea would be to devise a series of small refactorings, each of which being reasonably easy to verify, so that we end up in a state which has all operations to be guarded against concurrency happening in a single place. My hope is that the patch which then resolves the concurrency issue(s) is then a lot more obvious in how and why it helps. I.e. it not only does the right thing, it also does the Thing Right. :-)

**Comment by timblechmann** [comment]
#75 and #77 make this obsolete

**Pull request closed without merging**